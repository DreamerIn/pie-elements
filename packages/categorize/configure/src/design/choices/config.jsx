import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

import TextField from '@material-ui/core/TextField';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import { InputCheckbox } from '@pie-lib/config-ui';

const positionOptions = [
  {
    label: 'Above',
    value: 'above',
    secondaryText: 'Choices will be shown above categories'
  },
  {
    label: 'Below',
    value: 'below',
    secondaryText: 'Choices will be shown below categories'
  },
  {
    label: 'Left',
    value: 'left',
    secondaryText: 'Choices will be shown to the left of the categories'
  },
  {
    label: 'Right',
    value: 'right',
    secondaryText: 'Choices will be shown to the right of the categories'
  }
];

const withDefaults = o => ({
  label: '',
  columns: 2,
  position: 'above',
  shuffle: false,
  ...o
});
export class Config extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    config: PropTypes.object,
    onChange: PropTypes.func.isRequired,
    categoryCountIsOne: PropTypes.bool,
    onToggleCategoryCount: PropTypes.func
  };

  static defaultProps = {};

  state = {
    anchorEl: null
  };

  changeColumns = event => {
    const numberValue = parseInt(event.target.value, 10);

    if (numberValue && numberValue >= 1 && numberValue <= 4) {
      this.apply(config => (config.columns = numberValue));
    }
  };

  apply = fn => {
    const { onChange } = this.props;
    const config = withDefaults(this.props.config);
    fn(config);
    onChange(config);
  };

  changeLabel = event => {
    this.apply(config => (config.label = event.target.value));
  };

  toggleShuffle = () => {
    this.apply(config => (config.shuffle = !config.shuffle));
  };

  changePosition = position => {
    this.apply(config => (config.position = position.value));
    this.setState({ anchorEl: null });
  };

  handleClickPosition = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleMenuItemClick = (event, index) => {
    this.setState({ selectedIndex: index, anchorEl: null });
  };

  handleClose = () => {
    this.setState({ anchorEl: null });
  };

  render() {
    const { anchorEl } = this.state;
    const {
      classes,
      className,
      categoryCountIsOne,
      onToggleCategoryCount
    } = this.props;

    const config = withDefaults(this.props.config);

    const positionOption =
      positionOptions.find(option => option.value === config.position) ||
      positionOption[1];

    return (
      <div className={classNames(classes.config, className)}>
        <div className={classes.row}>
          <TextField
            label={'Choices per row'}
            type="number"
            inputProps={{
              min: 1,
              max: 4
            }}
            value={config.columns}
            onChange={this.changeColumns}
          />
          <InputCheckbox
            label={'Remove all tiles after placing'}
            checked={categoryCountIsOne}
            onChange={onToggleCategoryCount}
          />
          <InputCheckbox
            className={classes.shuffleCheckbox}
            label={'Shuffle'}
            value={config.shuffle}
            onChange={this.toggleShuffle}
          />
        </div>

        <div className={classes.configuration}>
          <TextField
            className={classes.label}
            InputLabelProps={{
              shrink: true
            }}
            label="Label"
            value={config.label}
            onChange={this.changeLabel}
          />
          <List component="nav">
            <ListItem
              button
              aria-haspopup="true"
              onClick={this.handleClickPosition}
            >
              <ListItemText
                primary="Choices Position"
                secondary={positionOption.secondaryText}
              />
            </ListItem>
          </List>
          <Menu
            id="lock-menu"
            anchorEl={anchorEl}
            open={Boolean(anchorEl)}
            onClose={this.handleClose}
          >
            {positionOptions.map(option => (
              <MenuItem
                key={option.value}
                selected={option.value === config.position}
                onClick={() => this.changePosition(option)}
              >
                {option.label}
              </MenuItem>
            ))}
          </Menu>
        </div>
      </div>
    );
  }
}
const styles = theme => ({
  configuration: {
    display: 'grid',
    alignItems: 'center',
    gridColumnGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridTemplateColumns: 'repeat(2, 1fr)',
    justifyContent: 'space-between',
    marginBottom: theme.spacing.unit
  },
  choiceConfig: {
    width: '100%',
    maxWidth: 360,
    backgroundColor: theme.palette.background.paper
  },
  row: {
    display: 'grid',
    gridColumnGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridTemplateColumns: 'repeat(2, 1fr)',
    justifyContent: 'space-between',
    marginTop: theme.spacing.unit,
    marginBottom: theme.spacing.unit
  },
  shuffleCheckbox: {
    gridColumnStart: 2,
    gridColumnEnd: 2
  }
});
export default withStyles(styles)(Config);
