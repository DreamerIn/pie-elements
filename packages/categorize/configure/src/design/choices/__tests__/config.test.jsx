import { shallow } from 'enzyme';
import React from 'react';

import { Config } from '../config';

describe('config', () => {
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
  });
  const wrapper = extras => {
    const props = { classes: {}, onChange, ...extras };
    return shallow(<Config {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      const w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('apply', () => {
      const assert = function() {
        const args = Array.prototype.slice.call(arguments);
        const method = args[0];
        const handler = args[args.length - 1];
        const fnArgs = args.slice(1, args.length - 1);
        it(`calls onChange for ${method}`, () => {
          const w = wrapper();
          w.instance()[method].apply(w.instance, fnArgs);
          handler();
        });
      };
      assert('changeColumns', {}, 4, () => {
        expect(onChange).toBeCalledWith({
          columns: 4,
          label: '',
          position: 'above',
          shuffle: false
        });
      });
      assert('changeLabel', { target: { value: 'foo' } }, () => {
        expect(onChange).toBeCalledWith({
          columns: 2,
          label: 'foo',
          position: 'above',
          shuffle: false
        });
      });
      assert('toggleShuffle', () => {
        expect(onChange).toBeCalledWith({
          columns: 2,
          label: '',
          position: 'above',
          shuffle: true
        });
      });
      assert('changePosition', 'below', () => {
        expect(onChange).toBeCalledWith({
          columns: 2,
          label: '',
          position: 'below',
          shuffle: false
        });
      });
    });
  });
});
