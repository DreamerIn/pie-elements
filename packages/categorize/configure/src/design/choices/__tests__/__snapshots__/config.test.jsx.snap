// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config snapshot renders 1`] = `
<div
  className=""
>
  <div>
    <TextField
      inputProps={
        Object {
          "max": 4,
          "min": 1,
        }
      }
      label="Choices per row"
      onChange={[Function]}
      required={false}
      select={false}
      type="number"
      value={2}
      variant="standard"
    />
    <Component
      label="Remove all tiles after placing"
      onChange={[Function]}
    />
    <Component
      label="Lock Choice Order"
      onChange={[Function]}
      value={true}
    />
  </div>
  <div>
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      label="Label"
      onChange={[Function]}
      required={false}
      select={false}
      value=""
      variant="standard"
    />
    <WithStyles(List)
      component="nav"
    >
      <WithStyles(ListItem)
        aria-haspopup="true"
        button={true}
        onClick={[Function]}
      >
        <WithStyles(ListItemText)
          primary="Choices Position"
          secondary="Choices will be shown below categories"
        />
      </WithStyles(ListItem)>
    </WithStyles(List)>
    <WithStyles(Menu)
      anchorEl={null}
      id="lock-menu"
      onClose={[Function]}
      open={false}
    >
      <WithStyles(MenuItem)
        key="above"
        onClick={[Function]}
        selected={false}
      >
        Above
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="below"
        onClick={[Function]}
        selected={true}
      >
        Below
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="left"
        onClick={[Function]}
        selected={false}
      >
        Left
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="right"
        onClick={[Function]}
        selected={false}
      >
        Right
      </WithStyles(MenuItem)>
    </WithStyles(Menu)>
  </div>
</div>
`;
